Refactor race condition reblog service spec

mjankowski opened the pull request:
This change tries to more narrowly create the scenario where a status being reblogged is discarded after validations have run but before the database commit occurs in the Status doing the reblogging.

After this spec improvement yesterday - https://github.com/mastodon/mastodon/pull/24516/files - I rebased my Rails 7 branch, and was able to get the spec passing by adding a `reload` to my validation approach - https://github.com/mastodon/mastodon/pull/24241/commits/40ff3cd6bbb551babde578fbee0715a83d94390e#diff-b8b0d989b8031e6690947f5a1117fa51873f60a213b202a7e97b15bd873c2f08R492

This further update to the spec creates an even narrower simulation of the race condition we want to handle, and importantly a) fails against my validation approach in my rails 7 branch, b) passes with the monkey patched insert methods in Status.

mjankowski committed:
Refactor race condition reblog service spec

This change tries to more narrowly create the scenario where a status
being reblogged is discarded after validations have run but before
the database commit occurs in the Status doing the reblogging.

ClearlyClaire reviewed (changes requested):
I like the idea but I'm uncomfortable with adding code to the class like this. I wonder if `rspec` provides better suited facilities for this. I'll investigate. In the meantime, I have made a few inline comments.

EDIT: I have not been able to find a better way so far.

mjankowski commented on spec/services/reblog_service_spec.rb:
```diff
@@ -35,13 +35,24 @@
-      # pretty close.
-      Status.where(id: status.id).update_all(deleted_at: Time.now.utc) # rubocop:disable Rails/SkipsModelValidations
+      Status.class_eval do
+        after_validation :discard_status
```
I think those run sequentially, but I can update to use before create.

ClearlyClaire replied:
Yes, they definitely do, it's just further along in the chain, making the test more robust to changes (e.g. if a `before_save` hook is added and deals with the issue, the test would still catch the narrower window)

mjankowski commented on spec/services/reblog_service_spec.rb:
```diff
@@ -35,13 +35,24 @@
+        after_validation :discard_status
+        def discard_status
+          Status
+            .where(text: 'race-condition-discard')
```
Good idea -- I initially did it this way because the added callback method doesn't have access to vars from the spec, but I can change to just use the actual reblog. Will do and push in a bit.

mjankowski commented:
> I like the idea but I'm uncomfortable with adding code to the class like this. I wonder if `rspec` provides better suited facilities for this. I'll investigate. In the meantime, I have made a few inline comments.

I also don't love it, and would be open to other approaches. Thus far this is the least bad thing I can think of.

I went down paths of:

- Pass global vars around between spec and class, which felt even worse
- Use multiple threads to try to get the timing order right for deletion, but this also felt worse and also involved modifying the class either in the spec or actual class code.

I think the removal of the callback from the chain in the example clean up phase makes this low risk, if a little odd.

mjankowski committed:
Adjust reblog service spec

- Use before_save instead of after_validation callback
- Use the actual ID of the reblog in the instance
- Undefine the callback method in example cleanup

mjankowski commented:
Pushed suggested changes.

ClearlyClaire commented on spec/services/reblog_service_spec.rb:
```diff
@@ -37,11 +37,22 @@
+
+    # Remove race condition simulating `discard_status` callback.
+    after do
+      Status._validation_callbacks.delete(:discard_status)
```
This needs to be changed now that the callback has changed:
```suggestion
      Status._save_callbacks.delete(:discard_status)
```

ClearlyClaire commented on spec/services/reblog_service_spec.rb:
```diff
@@ -37,11 +37,22 @@
+      Status.class_eval do
+        before_save :discard_status
+        def discard_status
+          Status
+            .where(id: reblog_of_id)
+            .update_all(deleted_at: Time.now.utc) # rubocop:disable Rails/SkipsModelValidations
```
This should check that `reblog_of_id` is not `nil`, and/or have another way to identify the expected status (e.g. what you did before)

mjankowski commented:
Pushed change to fix removal from save callback chain instead of validation chain, and restore the where condition to get just the status we want to discard.
